Reload without content blockers may not appear for extensions that use DNR

elijahsawyers opened the pull request:
#### 97e744cbfdca05fad698f53a482b5c00c3c38a58
<pre>
Reload without content blockers may not appear for extensions that use DNR
<a href="https://webkit.org/b/270552">https://webkit.org/b/270552</a>
<a href="https://rdar.apple.com/124033486">rdar://124033486</a>

Reviewed by Timothy Hatcher.

Clients need to be aware of whether or not a web extension context has any
content modification rules rules. This patch adds a new property to expose
that information.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext hasContentModificationRules]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasContentModificationRules):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: <a href="https://commits.webkit.org/275768@main">https://commits.webkit.org/275768@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0138b6b6e25b58ab579a8d961c919fac65c08c68

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/42742 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/21763 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/45143 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/45354 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38866 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/45048 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/25429 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/19128 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/35377 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/43315 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/18769 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36788 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/16336 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/16399 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/37848 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/798 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38912 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/38179 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/46862 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/17560 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/14478 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42102 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/19179 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/40730 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9544 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/19358 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18824 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

elijahsawyers commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -365,6 +369,8 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
  */
 - (BOOL)hasInjectedContentForURL:(NSURL *)url;
 
+@property (nonatomic, readonly) BOOL hasDynamicOrSessionDeclarativeNetRequestRules;
```
This needs a comment block.

xeenon replied:
This should be a more generic name, I would copy the one was have for `_WKWebExtension` with a `@ discussion `.

Something like:

```objc
/*!
 @abstract A boolean value indicating whether the extension includes rules used for content modification or blocking.
 @discussion This includes both static rules available in the extension's manifest and dynamic rules applied during a browsing session.
 */
@property (nonatomic, readonly) BOOL hasContentModificationRules;
```

We want the API to be as generic as possible so it can adapt as the underlying implementation might change as new web extension mechanisms are added.

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -31,6 +31,10 @@
 #import <WebKit/_WKWebExtensionTab.h>
 
+#if TARGET_OS_OSX
+#import <AppKit/AppKit.h>
```
a forward declaration with`@class NSEvent` would be better than the import for AppKit

xeenon replied:
Yep! I'd group it with `NSMenuItem` below.

```objc
#if TARGET_OS_IPHONE
@class UIMenuElement;
#else
@class NSEvent;
@class NSMenuItem;
#endif
```

kiaraarose commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -284,6 +284,7 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     const InjectedContentVector& injectedContents();
     bool hasInjectedContentForURL(const URL&);
     bool hasInjectedContent();
+    bool hasDynamicOrSessionDeclarativeNetRequestRules();
```
nit: new line above since it's not related to injectedContent

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3704,6 +3704,11 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
     });
 }
 
+bool WebExtensionContext::hasDynamicOrSessionDeclarativeNetRequestRules()
```
Ditto for this internal name. Should be `hasContentModificationRules()`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
```diff
@@ -3704,6 +3704,11 @@ void effectiveAppearanceDidChange(Inspector::ExtensionAppearance appearance) ove
 
+bool WebExtensionContext::hasDynamicOrSessionDeclarativeNetRequestRules()
+{
+    return !m_sessionRulesIDs.isEmpty() || !m_dynamicRulesIDs.isEmpty();
```
This should call `declarativeNetRequestEnabledRulesetCount() || …` to count the static rules.

kiaraarose replied:
i think the intention here is that this method is for dynamic and sessions rules only

kiaraarose replied:
discussed on slack and decided it was best to have a full proof method that considers the dynamic, session, and enabled static rules

elijahsawyers committed:
Reload without content blockers may not appear for extensions that use DNR
https://webkit.org/b/270552
rdar://124033486

Reviewed by Timothy Hatcher.

Clients need to be aware of whether or not a web extension context has any
content modification rules rules. This patch adds a new property to expose
that information.

* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext hasContentModificationRules]):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::hasContentModificationRules):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:

Canonical link: https://commits.webkit.org/275768@main
